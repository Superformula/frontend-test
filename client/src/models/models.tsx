export class DdItem { // class model for Dropdown menu items
  content: string;
  selected: boolean;
  
  constructor(
    content: string,
    selected: boolean,
  ) {
    this.content = content;
    this.selected = selected;
  }
}

export class Review { // class model for Reviews
  revId: string; // Unique review id
  resId: string; // Id of restaurant that a given Review belongs to
  name: string; 
  date: string;
  rating: number;
  review: string;
  image: string;

  constructor(
    revId: string,
    resId: string,
    name: string,
    date: string,
    rating: number,
    review: string,
    image: string,
  ) {
    this.revId = revId;
    this.resId = resId;
    this.name = name;
    this.date = date;
    this.rating = rating;
    this.review = review;
    this.image = image
  }
}

export class Restaurant { // class model for Restaurants
  resId: string;
  title: string;
  category: string[];
  isOpen: boolean;
  cost: string;
  avg_rating: number;
  img: string;
  coordinates: number[];
  totReviews: number;
  reviews: Review[];
  location: string;
  constructor(
    resId: string,
    title: string,
    category: string[],
    isOpen: boolean,
    cost: string,
    avg_rating: number,
    img: string,
    totReviews: number,
    location: string,
    coordinates: number[],
    reviews: Review[],
  ) {
    this.resId = resId;
    this.title = title;
    this.category = category;
    this.isOpen = isOpen;
    this.cost = cost;
    this.avg_rating = avg_rating;
    this.img = img;
    this.coordinates = coordinates;
    this.totReviews = totReviews;
    this.reviews = reviews;
    this.location = location;
  }
}

