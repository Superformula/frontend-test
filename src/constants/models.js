import PropTypes from 'prop-types';

const coordinatesShape = PropTypes.shape({
  latitude: PropTypes.number,
  longitude: PropTypes.number,
});

export const yelpBusinessShape = PropTypes.shape({
  alias: PropTypes.string,
  categories: PropTypes.arrayOf(PropTypes.shape({
    alias: PropTypes.string,
    title: PropTypes.string,
  })),
  coordinates: coordinatesShape,
  display_phone: PropTypes.string,
  distance: PropTypes.number,
  id: PropTypes.string,
  image_url: PropTypes.string,
  is_closed: PropTypes.bool,
  location: PropTypes.shape({
    address1: PropTypes.string,
    address2: PropTypes.string,
    address3: PropTypes.string,
    city: PropTypes.string,
    country: PropTypes.string,
    display_address: PropTypes.arrayOf(PropTypes.string),
    state: PropTypes.string,
    zip_code: PropTypes.string,
  }),
  name: PropTypes.string,
  phone: PropTypes.string,
  price: PropTypes.string,
  rating: PropTypes.number,
  review_count: PropTypes.number,
  transactions: PropTypes.arrayOf(PropTypes.any),
  url: PropTypes.string,
});

export const yelpResultsShape = {
  businesses: PropTypes.arrayOf(yelpBusinessShape),
  region: PropTypes.shape({
    center: coordinatesShape,
  }),
  total: PropTypes.number,
};
