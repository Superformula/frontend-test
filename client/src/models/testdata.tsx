// export {} // to make typescript happy

import { Restaurant, Review } from './models';
import { costArr, categoriesArr } from '../models/ddData'


const imgs = {
  panda: 'https://d38trduahtodj3.cloudfront.net/images.ashx?t=ig&rid=VisitHenrico&i=panda_express.png&w=330&h=330&cropbox=1&cropboxhpos=center&stf=1',
  car: 'https://vectorlogoseek.com/wp-content/uploads/2018/09/carrabbas-italian-grill-vector-logo.png',
  cope: 'https://copelandsofneworleans.com/wp-content/uploads/2018/07/CNO-Red-TAGLINE.png',
  dragos: 'https://dragosrestaurant.com/wp-content/uploads/2018/01/dragos-image-logo.png',
  redL: 'https://upload.wikimedia.org/wikipedia/en/b/b1/Redlobster_logocopia.png',
  outb: 'https://ncccc.org/wp-content/uploads/2018/09/Outback-Steakhouse-logo.png',
  mcDon: 'https://news.mcdonalds.com/system/files-encrypted/styles/nir_media_item_grid_thumbnail/encrypt/nasdaq_kms/media_gallery/thumbnail/2019-04/The%20Token.png?itok=yRL6WXDB',
  fel: 'https://baltimorefishbowl.com/wp-content/uploads/2019/01/Felipes-logo-e1546464726282.jpg',
  chi: 'https://www.nrn.com/sites/nrn.com/files/uploads/2016/04/chipotlelogopromo_0.jpg',
  pho: 'https://www.menusie.com/uploads/5/2/9/9/52994693/pho-bistro-banner_orig.png',
};

const sampleReview = `Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt 
ut labore et dolore magna aliqua. Turpis tincidunt id aliquet risus feugiat in ante metus dictum. 
Suspendisse ultrices gravida dictum fusce. Aliquet nec ullamcorper sit amet risus. At consectetur lorem donec 
massa sapien faucibus et molestie. Amet justo donec enim diam vulputate ut pharetra sit. Cursus in hac habitasse 
platea dictumst. Aenean sed adipiscing diam donec adipiscing tristique risus nec. Dolor purus non enim praesent 
elementum facilisis leo. Eget aliquet nibh praesent tristique magna sit amet purus gravida. Maecenas accumsan 
lacus vel facilisis volutpat. Volutpat maecenas volutpat blandit aliquam etiam. Egestas fringilla phasellus 
faucibus scelerisque eleifend donec pretium vulputate. Egestas tellus rutrum tellus pellentesque eu tincidunt 
tortor aliquam nulla. Amet cursus sit amet dictum sit amet. Nisi porta lorem mollis aliquam ut. Dignissim diam 
quis enim lobortis scelerisque fermentum dui. Ipsum dolor sit amet consectetur adipiscing elit ut. Tellus molestie 
nunc non blandit massa enim nec. Et malesuada fames ac turpis egestas maecenas.`

const carReviews = [
  new Review('111', '1', 'Jon', '08/31/2019', 4, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
  new Review('112', '1', 'Sarah', '08/31/2019', 5, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
  new Review('113', '1', 'Imani', '08/31/2019', 4, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
  new Review('114', '1', 'Michael', '08/31/2019', 5, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
];

const copeReviews = [
  new Review('211', '2', 'Jon', '08/31/2019', 4, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
  new Review('212', '2', 'Sarah', '08/31/2019', 5, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
  new Review('213', '2', 'Imani', '08/31/2019', 5, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
  new Review('214', '2', 'Michael', '08/31/2019', 4, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
];

const dragReviews = [
  new Review('311', '3', 'Jon', '08/31/2019', 4, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
  new Review('312', '3', 'Sarah', '08/31/2019', 3, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
  new Review('313', '3', 'Imani', '08/31/2019', 5, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
  new Review('314', '3', 'Michael', '08/31/2019', 4, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
];

const redReviews = [
  new Review('411', '4', 'Jon', '08/31/2019', 4, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
  new Review('412', '4', 'Sarah', '08/31/2019', 3, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
  new Review('413', '4', 'Imani', '08/31/2019', 3, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
  new Review('414', '4', 'Michael', '08/31/2019', 4, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
];

const outReviews = [
  new Review('511', '5', 'Jon', '08/31/2019', 4, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
  new Review('512', '5', 'Sarah', '08/31/2019', 4, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
  new Review('513', '5', 'Imani', '08/31/2019', 4, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
  new Review('514', '5', 'Michael', '08/31/2019', 4, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
]

const pandaReviews = [
  new Review('611', '6', 'Jon', '08/31/2019', 4, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
  new Review('612', '6', 'Sarah', '08/31/2019', 3, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
  new Review('613', '6', 'Imani', '08/31/2019', 3, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
  new Review('614', '6', 'Michael', '08/31/2019', 4, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
];

const mcReviews = [
  new Review('711', '7', 'Jon', '08/31/2019', 1, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
  new Review('712', '7', 'Sarah', '08/31/2019', 2, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
  new Review('713', '7', 'Imani', '08/31/2019', 3, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
  new Review('714', '7', 'Michael', '08/31/2019', 4, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
];

const felReviews = [
  new Review('811', '8', 'Jon', '08/31/2019', 3, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
  new Review('812', '8', 'Sarah', '08/31/2019', 4, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
  new Review('813', '8', 'Imani', '08/31/2019', 3, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
  new Review('814', '8', 'Michael', '08/31/2019', 5, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
];

const chiReviews = [
  new Review('911', '9', 'Jon', '08/31/2019', 3, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
  new Review('912', '9', 'Sarah', '08/31/2019', 4, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
  new Review('913', '9', 'Imani', '08/31/2019', 5, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
  new Review('914', '9', 'Michael', '08/31/2019', 5, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
];

const phoReviews = [
  new Review('1011', '10', 'Jon', '08/31/2019', 5, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
  new Review('1012', '10', 'Sarah', '08/31/2019', 4, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
  new Review('1013', '10', 'Imani', '08/31/2019', 4, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
  new Review('1014', '10', 'Michael', '08/31/2019', 4, sampleReview, 'https://img1.looper.com/img/gallery/mclovin-from-superbad-looks-totally-different-today/intro-1535553409.jpg'),
];


export const exRestaurants = [
  new Restaurant('1', 'Carrabas', [categoriesArr[1].content], true, costArr[3].content, 4, imgs.car, 3425, 'Pete Rose Blvd', [55.3426, -46.43524], carReviews),
  new Restaurant('2', 'Copelands', [categoriesArr[1].content, "Italian", "Delicious"], false, costArr[4].content, 4, imgs.cope, 23456, 'Pete Rose Blvd', [55.3426, -46.43524], copeReviews),
  new Restaurant('3', 'Dragos', [categoriesArr[2].content], false, costArr[4].content, 4, imgs.dragos, 4591, 'Pete Rose Blvd', [55.3426, -46.43524], dragReviews),
  new Restaurant('4', 'Red Lobster', [categoriesArr[2].content], true, costArr[3].content, 4, imgs.redL, 2346, 'Pete Rose Blvd', [55.3426, -46.43524], redReviews),
  new Restaurant('5', 'Outback', [categoriesArr[3].content], true, costArr[4].content, 4, imgs.outb, 324566, 'Pete Rose Blvd', [55.3426, -46.43524], outReviews),
  new Restaurant('6', 'Panda Express', [categoriesArr[4].content], false, costArr[1].content, 4, imgs.panda, 34571, 'Pete Rose Blvd', [55.3426, -46.43524], pandaReviews),
  new Restaurant('7', 'McDonalds', [categoriesArr[5].content], false, costArr[1].content, 4, imgs.mcDon, 43735, 'Pete Rose Blvd', [55.3426, -46.43524], mcReviews),
  new Restaurant('8', 'Felipe\'s', [categoriesArr[6].content], false, costArr[2].content, 4, imgs.fel, 134785, 'Pete Rose Blvd', [55.3426, -46.43524], felReviews),
  new Restaurant('9', 'Chipotle', [categoriesArr[6].content], false, costArr[3].content, 4, imgs.chi, 364545, 'Pete Rose Blvd', [55.3426, -46.43524], chiReviews),
  new Restaurant('10', 'Pho Bistro', [categoriesArr[7].content], true, costArr[4].content, 4, imgs.pho, 345672, 'Pete Rose Blvd', [55.3426, -46.43524], phoReviews),
];