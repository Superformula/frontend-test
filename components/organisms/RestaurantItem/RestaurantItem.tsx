import classNames from 'classnames';
import React, { FC } from 'react';
import Skeleton from 'react-loading-skeleton';

import {
  Button,
  ButtonProps,
  Cover,
  Typography,
} from '~/components/atoms';

import {
  CategoryStatus,
  CategoryStatusProps,
  OpenStatus,
  OpenStatusProps,
  Rating,
  RatingProps,
} from '~/components/molecules';

import styles from './RestaurantItem.module.scss';

export interface RestaurantItemProps {
  category: string;
  imageUrl: string;
  loading?: boolean;
  price: CategoryStatusProps['price'];
  rating: RatingProps['value'];
  status: OpenStatusProps['status'];
  title: string;
  onClick: ButtonProps['onClick'];
}

export const RestaurantItem: FC<RestaurantItemProps> = ({
  category, imageUrl, price, rating, status, title, onClick, loading,
}) => {
  const elementClass = classNames({
    [styles.item]: true,
  });

  const coverClass = classNames({
    [styles['item-cover']]: true,
  });

  const headlineClass = classNames({
    [styles['item-headline']]: true,
  });

  const ratingClass = classNames({
    [styles['item-rating']]: true,
  });

  const statusClass = classNames({
    [styles['item-status']]: true,
  });

  const buttonClass = classNames({
    [styles['item-button']]: true,
  });

  return (
    <div className={elementClass}>
      <div className={coverClass}>
        {loading && <Skeleton height={228} />}
        {!loading && <Cover src={imageUrl} alt="cover" />}
      </div>

      <div className={headlineClass}>
        {loading && <Skeleton height={25} />}
        {!loading && <Typography variant="headline">{title}</Typography>}
      </div>

      <div className={ratingClass}>
        {loading && <Skeleton height={21.3} />}
        {!loading && <Rating value={rating} />}
      </div>

      <div className={statusClass}>
        {loading && <Skeleton height={16} />}

        {
          !loading && (
            <div>
              <CategoryStatus category={category} price={price} />
              <OpenStatus status={status} />
            </div>
          )
        }
      </div>

      <div className={buttonClass}>
        {loading && <Skeleton height={48} />}

        {!loading && <Button size="full-width" color="primary" onClick={onClick}>Learn More</Button>}
      </div>
    </div>
  );
};
