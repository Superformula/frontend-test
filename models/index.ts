export * from './Business';
export * from './Category';
export * from './Review';
