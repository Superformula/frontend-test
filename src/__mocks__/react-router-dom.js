import React from 'react'

const mockRouter = require('react-router-dom')

// Just render plain div with its children
mockRouter.BrowserRouter = ({ children }) => <div>{children}</div>

module.exports = mockRouter
