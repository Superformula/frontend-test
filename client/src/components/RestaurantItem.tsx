import * as React from 'react'
import { useState } from 'react'
import { Link } from 'react-router-dom'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { Redirect } from 'react-router-dom';

import { Restaurant } from '../models/models'
import StarRating from './StarRating'

interface rListItemProps {
  R: Restaurant,
}

const RestaurantItem = ({ R }: rListItemProps) => {
  
  const [clicked, changeClick] = useState(false) // Manage per item state with react hooks

  const learnMoreClicked = () => {
    changeClick(true); // change state of clicked to true when clicked
  }

  // Send to reviews page when clicked is true
  return (clicked) ? <Redirect to={`/reviews/${R.resId}`}></Redirect> : (
    <li className="card-item">
      <div className="card">

        <div className="card-media">
          <img src={ R.img }/>
        </div>

        <div className="card-content">
          
          <div className="card-title">{ R.title }</div>

          <div className="card-rating">
            <StarRating rating={ R.avg_rating } />
          </div>

          <div className="card-text">
            <div className="text-container">
              <div className="meta-left">
                { R.category.map((str, ind) => <div key={ind}>{str}</div>) }
              </div>
              <div className="meta-center">
                { `Â· ${R.cost}`}
              </div>
              <div className="meta-right">
                {R.isOpen ? <FontAwesomeIcon icon="circle" color="#00E8A4" /> : <FontAwesomeIcon icon="circle" color="#FF3548"/> }
                { R.isOpen ? '  OPEN NOW' : '  CLOSED' }
              </div>
            </div>
          </div>

          <button className="card-button btn-block" onClick={() => learnMoreClicked()}> 
            <Link to={`/reviews/${R.resId}`} style={{ color: '#FFF' }}>LEARN MORE</Link>
          </button>
        
        </div>
      </div>
    </li>
  )
}

const MemRestaurantItem = React.memo(RestaurantItem)

export default MemRestaurantItem
