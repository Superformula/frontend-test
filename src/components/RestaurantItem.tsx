import * as React from 'react'
import { Link } from 'react-router-dom'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { Restaurant } from '../models/models'
import StarRating from './StarRating'

interface rListItemProps {
  R: Restaurant,
}

export default function RestaurantItem({ R }: rListItemProps) {
  return (
    <li className="card-item">
      <div className="card">

        <div className="card-media">
          <img src={ R.img }/>
        </div>

        <div className="card-content">
          
          <div className="card-title">{ R.title }</div>

          <div className="rating">
            <StarRating rating={ R.avg_rating } />
          </div>

          <div className="card-text">
            <div className="text-container">
              <div className="meta-left">
                { R.category.map((str, ind) => <p key={ind}>{str}</p>) } 
                { `Â· ${R.cost}`}
              </div>
              <div className="meta-right">
                { R.isOpen ? <FontAwesomeIcon icon="circle" color="green" /> : <FontAwesomeIcon icon="circle" color="red"/> }
                { R.isOpen ? '  OPEN NOW' : '  CLOSED' }
              </div>
            </div>
          </div>

          <button className="btn btn-block" onClick={() => {}}> 
            <Link to={`/reviews/${R.resId}`}>LEARN MORE</Link>
          </button>
        
        </div>
      </div>
    </li>
  )
}
