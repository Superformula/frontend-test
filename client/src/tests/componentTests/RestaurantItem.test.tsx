import * as React from 'react'
import * as renderer from 'react-test-renderer';
import { Provider } from 'react-redux'
import { faStar, faAngleUp, faAngleDown, faCircle, faCheckCircle, faStarHalfAlt } from '@fortawesome/free-solid-svg-icons'
import { faStar as farStar, faCircle as farCircle, faDotCircle } from '@fortawesome/free-regular-svg-icons'
import { library } from '@fortawesome/fontawesome-svg-core'

library.add(farStar, faStar, faAngleUp, faAngleDown, faCircle, farCircle, faDotCircle, faCheckCircle, faStarHalfAlt)

import RestaurantItem from '../../components/RestaurantItem'
import store from '../helpers'
import { Restaurant } from '../../models/models';
import { exRestaurants } from '../../models/testdata'
import { MemoryRouter } from 'react-router-dom'

interface rListItemProps {
  R: Restaurant,
}

const preTest = (props: rListItemProps) => {
  const component: any = renderer.create(
    <Provider store={store}>
      <MemoryRouter>
        <RestaurantItem {...props} />
      </MemoryRouter>
    </Provider>
  ).toJSON()
  return component;
}

describe('RestaurantItem Component', () => {

  describe('Initial render', () => {
    let component: any;
    
    describe('Static info', () => {
      const basicRest: rListItemProps = {
        R: exRestaurants[0]
      } 
    
      beforeEach(() => {  // assign necessary depencies before each test
        component = preTest(basicRest)
      })
    
      it('Should render without errors', () => {
        expect(component).toMatchSnapshot(); // checks snapshot of loaded JSON data from component
      })
    
      it('Should render correct Restaurant Image', () => {
        // checks if the image src property matches the string passed onto the Restaurant passed in
        const imageCheck: boolean = component.children[0].children[0].children[0].props.src === basicRest.R.img
        expect(imageCheck).toBe(true)
      })

    })
    
    describe('Dynamic Info', () => {
      
      it('Should render multiple Categories', () => {
        const multipleCatsRest: rListItemProps = {
          R: exRestaurants[1]
        } 
        component = preTest(multipleCatsRest)
        // checks if the length of categories is the same as the length on the passed restaurant
        const categoryCheck: boolean = component.children[0].children[1].children[2].children[0].children[0].children.length === multipleCatsRest.R.category.length
        expect(categoryCheck).toBe(true)
      })

      it('Should render correct colors based on open status', () => {
        const openRest: rListItemProps = {
          R: exRestaurants[4]
        } 
        const closedRest: rListItemProps = {
          R: exRestaurants[5]
        } 
        const openComponent = preTest(openRest)
        const closedComponent = preTest(closedRest)
        // check colors stored on open and closed components to see if they match the correct colors
        const openCheck: boolean = openComponent.children[0].children[1].children[2].children[0].children[2].children[0].props.color === '#00E8A4'
        const closedCheck: boolean = closedComponent.children[0].children[1].children[2].children[0].children[2].children[0].props.color === '#FF3548'
        
        expect(openCheck && closedCheck).toBe(true)
      })
  
      it('Should render correct open status text', () => {
        const openRest: rListItemProps = {
          R: exRestaurants[4]
        }
        const closedRest: rListItemProps = {
          R: exRestaurants[5]
        }
        const openComponent = preTest(openRest)
        const closedComponent = preTest(closedRest)
        // check if correct text rendered based on close boolean

        const openTextCheck: boolean = openComponent.children[0].children[1].children[2].children[0].children[2].children[1] === '  OPEN NOW'
        const closedTextCheck: boolean = closedComponent.children[0].children[1].children[2].children[0].children[2].children[1] === '  CLOSED'

        expect(openTextCheck && closedTextCheck).toBe(true)
      })
    })

  })
})