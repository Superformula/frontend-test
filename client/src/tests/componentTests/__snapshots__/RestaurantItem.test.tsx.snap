// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RestaurantItem Component Initial render Static info Should render without errors 1`] = `
<li
  className="card-item"
>
  <div
    className="card"
  >
    <div
      className="card-media"
    >
      <img
        src="https://vectorlogoseek.com/wp-content/uploads/2018/09/carrabbas-italian-grill-vector-logo.png"
      />
    </div>
    <div
      className="card-content"
    >
      <div
        className="card-title"
      >
        Carrabas
      </div>
      <div
        className="card-rating"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-star fa-w-18 fa-1x "
          color="#002B56"
          data-icon="star"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 576 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-star fa-w-18 fa-1x "
          color="#002B56"
          data-icon="star"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 576 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-star fa-w-18 fa-1x "
          color="#002B56"
          data-icon="star"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 576 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-star fa-w-18 fa-1x "
          color="#002B56"
          data-icon="star"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 576 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-star fa-w-18 fa-1x "
          color="#002B56"
          data-icon="star"
          data-prefix="far"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 576 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M528.1 171.5L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6zM388.6 312.3l23.7 138.4L288 385.4l-124.3 65.3 23.7-138.4-100.6-98 139-20.2 62.2-126 62.2 126 139 20.2-100.6 98z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </div>
      <div
        className="card-text"
      >
        <div
          className="text-container"
        >
          <div
            className="meta-left"
          >
            <div>
              Italian
            </div>
          </div>
          <div
            className="meta-center"
          >
            Â· $$$
          </div>
          <div
            className="meta-right"
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-circle fa-w-16 "
              color="#00E8A4"
              data-icon="circle"
              data-prefix="fas"
              focusable="false"
              role="img"
              style={Object {}}
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
              OPEN NOW
          </div>
        </div>
      </div>
      <button
        className="card-button btn-block"
        onClick={[Function]}
      >
        <a
          href="/reviews/1"
          onClick={[Function]}
          style={
            Object {
              "color": "#FFF",
            }
          }
        >
          LEARN MORE
        </a>
      </button>
    </div>
  </div>
</li>
`;
